### ISSUE TITLE
Adding Jira tasks in fast succession In Sprint Poker Scope stage is unreliable

### ISSUE THREAD

**Issue by Dschoordsch** [ParabolInc/parabol#5554]
### How to reproduce
Run a Sprint Poker and import new tasks from Jira during Scope stage in fast succession

### Expected
All tasks are added to the Estimate stage

### Actual
Some tasks seem randomly be deselected

![Adding Jira tasks in fast succession](https://user-images.githubusercontent.com/7331043/138254754-26252d01-120d-45be-8c28-113c28511a16.gif)

**Estimated effort:** 4 hours

**Comment by mattkrick** [comment]
oh this is great! 

i thought Igor's BP would have fixed this-- it uses a locking queue mechanism to make sure that events happen in the order they are received: https://github.com/ParabolInc/parabol/blob/17ab2c1cc49a4e659ddbd8eef50dfe741a286d2b/packages/server/graphql/mutations/updatePokerScope.ts

As a way forward, I'd propose the following:
- see why the redis lock queue doesn't fix this.
- if it's still a mystery, we can use robust messaging on mutations by switching to `syn: true` here https://github.com/ParabolInc/parabol/blob/78bb1a39ab083bebb2550a524a5d5efd526acff8/packages/server/socketHandlers/handleMessage.ts

**Comment by igorlesnenko** [comment]
@mattkrick I have a theory:

- We load a meeting from the database **before** setting a lock
- Then set a lock
- Then make an update using this previously loaded meeting

So two requests that come at the same time will load same meeting from the database.
And the second request after waiting in the queue will operate with this outdated meeting.

Solution is to load a meeting after setting a lock.

**Comment by nickoferrall** [comment]
It looks like @tianrunhe shotgunned this one in Slack, so I've added you to the issue so no one else steals it from you

**Comment by tianrunhe** [comment]
I see two issues:

1.  As Igor stated, we load meeting object before setting the lock.
2.  Two mutations could be fired up close enough for the same item, resulting the slightly latter one gets ignored per [this line](https://github.com/ParabolInc/parabol/blob/76c512e3f8f3662eeee71e092dc2955af2422bc7/packages/client/components/JiraScopingSearchResultItem.tsx#L59) as the first mutation is still submitting.

**Comment by tianrunhe** [comment]
I wasn't able to reproduce with multiple issues so I focused on the fast clicking behavior for one issue.

On `master` branch, if you fast click twice on an issue, hoping it to be unchecked, that's not the case:
![2021-12-09 15 59 34](https://user-images.githubusercontent.com/1879975/145494826-03defc20-407b-43a9-8017-670c25823945.gif)
This is because the second point I mentioned in previous comment. First click triggered the mutation for adding the issue to the phase. Second click immediately fires, but since the first mutation was still submitting, second click wouldn't do anything (because [this line](https://github.com/ParabolInc/parabol/blob/master/packages/client/components/JiraScopingSearchResultItem.tsx#L59)).
So what if we remove the `submitting` condition check? Now, both of the two mutations will fire. But here's what happens afterward:
![2021-12-09 16 01 20](https://user-images.githubusercontent.com/1879975/145495284-33006526-0588-44dd-9669-64a4743da274.gif)
This is because that although the mutation for unchecking fired slightly later, it returned faster than the first mutation. When the client receives the response from the first mutation, it marks the checkbox as checked again. This can be seen from the receiving timestamp in the websocket:
<img width="842" alt="Screen Shot 2021-12-09 at 3 29 56 PM" src="https://user-images.githubusercontent.com/1879975/145495565-7b06ec38-ea62-42b5-b388-b23365654538.png">
So, to really fix this, we need to enable reliable message queue for mutation as well (it's only enabled for subscription today).

**Comment by tianrunhe** [comment]
In the spirit of making small changes and moving fast, we propose a simpler solution: when user selects an issue, we just disable that checkbox until the mutation is finished, i.e., what GitHub is doing today on to-do checkboxes.

**Comment by mattkrick** [comment]
reliable messaging can be enabled in 1 LOC. any reason not to do it?

**Comment by tianrunhe** [comment]
### Update:

  
I turned on reliable messaging for mutation [here](https://github.com/ParabolInc/parabol/commit/70426fa7b2c5d9c92d4b26facad183608b0b02e8) but the issue still exists. A quick & reliable reproduction step is to quickly double click an issue:

![](https://user-images.githubusercontent.com/1879975/148110835-a5a1cabf-6a6c-4e56-bc9d-9884e82d4fdf.gif)

Looking at response from WebSocket & logging, it appears that the ADD mutation fired before DELETE. ADD mutation locks the meeting object and it's only when ADD finishes, DELETE grabs the lock and starts processing it. Everything works as expected so far.

> 2022-01-04T19:06:53.681Z :: ADD action is gonna acquire lock for meeting 95XVaRtL9K  
> 2022-01-04T19:06:53.681Z :: ADD action is gonna wait for lock for meeting 95XVaRtL9K  
> 2022-01-04T19:06:53.716Z :: ADD action has completed waiting for lock for meeting 95XVaRtL9K  
> 2022-01-04T19:06:53.987Z :: DELETE action is gonna acquire lock for meeting 95XVaRtL9K  
> 2022-01-04T19:06:53.987Z :: DELETE action is gonna wait for lock for meeting 95XVaRtL9K  
> 2022-01-04T19:06:54.179Z :: ADD action is gonna release lock for meeting 95XVaRtL9K  
> 2022-01-04T19:06:54.337Z :: DELETE action has completed waiting for lock for meeting 95XVaRtL9K  
> 2022-01-04T19:06:54.462Z :: DELETE action is gonna release lock for meeting 95XVaRtL9K

However the mystery is this: even though the ADD action finishes at 19:06:54.179, the GQL message sent to the client won't happen until 19:06:55.340. Whereas for the DELETE action, although it finishes a bit later at 19:06:54.462, the GQL message was sent right after it at 19:06:54.497 and well before the ADD action:

> 2022-01-04T19:06:54.497Z :: Going to send message for 10 (the DELETE action)  
> 2022-01-04T19:06:55.340Z :: Going to send message for 9 (the ADD action)

Therefore it doesn't matter if the client has reliable messaging turned on or not, it's the server who did not send the response in the order of incoming request.

@mattkrick baton passed to you 💪

**Comment by jordanh** [comment]
**Story Points: 13**
  [See the discussion](https://action.parabol.co/meet/8Kr615RAwE/estimate/9) in Sprint Poker #​​9

  *Powered by [Parabol](https://www.parabol.co/resources/sprint-poker-meetings)*

**Comment by BartoszJarocki** [comment]
I was able to reproduce the issue only a couple of times, and since Bruce had already investigated the ordering of mutations, my assumption was it could be performance-related.

I noticed a few things. In `UpdatePokerScopeMutation` we only use `optimisticUpdater`; there’s no `updater` implementation which means there could be a brief moment where the UI could be out of sync with actual data. 

To be specific, this is what’s happening when someone clicks the checkbox.
1. `UpdatePokerScopeMutation` is fired, network request is being made.
2. `optimisticUpdater` is executed, it updates the Relay store, we see the correctly updated UI.
3. Network response is received; whatever was done in Relay store by `optimisticUpdater` is being reverted.
4. Whatever was received in the network response is being normalized and saved to store by Relay.
5. All the components that use the updated data are being updated; now, we see an actual state in the UI.

A potential issue with the flow above is that Relay can take some time to normalize the network response, save it to the store and update all the components (and it’ll be all happening when our optimistic update is already reverted if I’m not mistaken). 

I checked if my assumptions were correct and did the following.
1. I removed fetching any data as a result of the `UpdatePokerScopeMutation` to see if Relay correctly reverts whatever was done by the optimistic updater
![1](https://user-images.githubusercontent.com/1017620/148241593-745aa092-6068-495a-a71d-64856dc05769.gif)

2. Just to make sure, I removed the optimistic updater and made sure the checkbox won’t change state without that.
![2](https://user-images.githubusercontent.com/1017620/148241611-9805fcab-a46d-49c2-83f3-756ccb0813e5.gif)

3. I added back fetching data in mutation response to check if Relay correctly normalizes the response and updates all the related components (even thought, there's no optimistic updater)
![3](https://user-images.githubusercontent.com/1017620/148241637-88c94ad9-47da-45dd-8526-c7823c886a49.gif)

4. I checked the profiler, and indeed processing the data after successful `UpdatePokerScopeMutation` takes ~200ms
 ![Screenshot 2022-01-05 at 16 08 25](https://user-images.githubusercontent.com/1017620/148241686-e7b4c5fe-7a8f-4e3f-be0f-605120de263b.png)

5. I added some logs to see how much does it take to execute the mutation on the server-side and client-side 
![Screenshot 2022-01-05 at 16 07 38](https://user-images.githubusercontent.com/1017620/148241816-6721cd77-7575-46dc-8b2f-219de1ab5274.png)

To summarize, my assumption is that adding a proper updater function would help with state being out of sync with real data in some cases (I think the faster the machine, the harder is to reproduce the issue). 

Also, even though we update the UI optimistically, we do not allow to click the checkbox again until the mutation succeeds which might take up to 1s. As everything >60ms looks like a lag, we could improve the UX here too. Possible would be to keep the checkbox state within the component and just sync that state with the server (but use the local state as a source of truth). Anyway, @mattkrick, @tianrunhe please let me know what do you think and if I should continue to work on this.

**Comment by mattkrick** [comment]
an updater function is _in addition_ to the normalization that relay already performs.
if you completely remove the optimistic updater, you'll see that the values still update as expected, it just takes a moment for the check boxes to get checked.

your intuition of believing it's something to do with the relay publishing queue is an interesting one, and is in fact a large problem within relay (one that i've been very vocal about in the OSS community). relay runs all server updates & then all optimistic updaters IIRC (could be reversed, either way, it doesn't occur in linear time & doesn't follow lamport's happened-before relation). if you think you can come up with a fix in the next few hours by all means keep digging, otherwise don't feel bad about giving this one up & i'll set up a loom & try to solve it in real time so we can all learn together

**Comment by BartoszJarocki** [comment]
oh yeah, in 3. I double checked if it updates the values without optimistic updater, and yeah, like you said it just takes a little bit more time. 

Regarding updater, I thought that having providing the updater might speed the process up but you're probably right. I'll spend a bit more time and then move on, unless I figure out something worth trying. 

And yes, I'd love to see the loom!

**Comment by tianrunhe** [comment]
@BartoszJarocki interesting investigation! It's such a mystery, isn't it? Just a quick note on:   
> Also, even though we update the UI optimistically, we do not allow to click the checkbox again until the mutation succeeds which might take up to 1s
This was just a blue tape I put on for the issue and it's not what's in master today. Matt and I agreed it's not the desired behavior.
